async_processor/kafka_pub_sub.py

```diff
@@ -0,0 +1,116 @@
+from contextlib import asynccontextmanager
+from typing import AsyncIterator, Optional
+
+from kafka import KafkaConsumer, KafkaProducer
+from starlette.concurrency import run_in_threadpool
+
+from async_processor.logger import logger
+from async_processor.types import (
+    Input,
+    InputFetchAckFailure,
+    InputMessageFetchFailure,
+    KafkaInputConfig,
+    KafkaOutputConfig,
+    Output,
+)
+
+
+class KafkaInput(Input):
+    def __init__(self, config: KafkaInputConfig):
+        self._bootstrap_servers = config.bootstrap_servers
+        self._topic_name = config.topic_name
+        self._auth = config.auth
+        self._consumer_group_id = config.consumer_group_id
+        self.wait_time_seconds = config.wait_time_seconds
+        self._consumer = KafkaConsumer(
+            self._topic_name,
+            bootstrap_servers=self._bootstrap_servers,
+            group_id=config.consumer_group_id,
+            enable_auto_commit=False,
+            **(
+                {
+                    "sasl_plain_username": config.auth.username,
+                    "sasl_plain_password": config.auth.password,
+                    "security_protocol": "SASL_SSL",
+                    "sasl_mechanism": "PLAIN",
+                }
+                if config.auth
+                else {}
+            ),
+        )
+        self._producer = KafkaProducer(
+            bootstrap_servers=self._bootstrap_servers,
+            **(
+                {
+                    "sasl_plain_username": config.auth.username,
+                    "sasl_plain_password": config.auth.password,
+                    "security_protocol": "SASL_SSL",
+                    "sasl_mechanism": "PLAIN",
+                }
+                if config.auth
+                else {}
+            ),
+        )
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[str]]:
+        try:
+            consumer_map = await run_in_threadpool(
+                self._consumer.poll,
+                timeout_ms=self.wait_time_seconds * 1000,
+                max_records=1,
+            )
+        except Exception as ex:
+            raise InputMessageFetchFailure() from ex
+
+        if len(consumer_map) == 0:
```

Comment:
Is it possible that this dictionary is non-empty,

```
        for _, msgs in consumer_map.items():
            for msg in msgs:
```
But `msgs` is empty.

Reply:
no, i tested that, so if msgs is empty in that case map is also empty